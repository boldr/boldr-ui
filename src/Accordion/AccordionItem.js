/* @flow */
import React from 'react';
import type { Node } from 'react';
import uniqueId from 'lodash.uniqueid';

type Props = {
  isExpanded?: boolean,
  isAccordion?: boolean,
  onClick: Function,
  children: Array<Node>,
};

type State = {
  itemUuid: string,
};

class AccordionItem extends React.Component<Props, State> {
  static defaultProps = {
    isAccordion: true,
    isExpanded: false,
    onClick: () => {},
  };
  constructor() {
    super();

    this.state = {
      itemUuid: uniqueId(),
    };
  }
  props: Props;
  renderChildren = () => {
    const { isAccordion, isExpanded, onClick, children } = this.props;
    const { itemUuid } = this.state;

    return React.Children.map(children, item => {
      const itemProps = {};

      if (item.type.accordionElementName === 'AccordionItemTitle') {
        itemProps.isExpanded = isExpanded;
        itemProps.key = 'title';
        itemProps.id = `boldr-accordion__title-${itemUuid}`;
        itemProps.ariaControls = `boldr-accordion__body-${itemUuid}`;
        itemProps.onClick = onClick;
        itemProps.role = isAccordion ? 'tab' : 'button';

        return React.cloneElement(item, itemProps);
      } else if (item.type.accordionElementName === 'AccordionItemBody') {
        itemProps.isExpanded = isExpanded;
        itemProps.key = 'body';
        itemProps.id = `boldr-accordion__body-${itemUuid}`;
        itemProps.role = isAccordion ? 'tabpanel' : '';

        return React.cloneElement(item, itemProps);
      }

      return item;
    });
  };

  render() {
    return <div className="boldr-accordion__item">{this.renderChildren()}</div>;
  }
}

export default AccordionItem;
