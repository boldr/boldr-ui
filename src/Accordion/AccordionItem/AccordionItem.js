/* @flow */
import React, { Component } from 'react';
import uuid from 'uuid';

type Props = {
  expanded: boolean,
  accordion: boolean,
  onClick: () => void,
  children: ReactChildren,
};

class AccordionItem extends Component {
  static defaultProps = {
    accordion: true,
    expanded: false,
    onClick: () => {},
  };
  constructor(props) {
    super(props);
    this.state = {
      itemUuid: uuid.v4(),
    };
    this.renderChildren = this.renderChildren.bind(this);
  }
  props: Props;
  renderChildren() {
    const { accordion, expanded, onClick, children } = this.props;
    const { itemUuid } = this.state;

    return React.Children.map(children, item => {
      const itemProps = {
        expanded,
      };

      if (item.type.accordionElementName === 'AccordionItemTitle') {
        itemProps.key = 'title';
        itemProps.id = `boldrui-accordion__title-${itemUuid}`;
        itemProps.ariaControls = `boldrui-accordion__body-${itemUuid}`;
        itemProps.onClick = onClick;
        itemProps.role = accordion ? 'tab' : 'button';
      } else if (item.type.accordionElementName === 'AccordionItemBody') {
        itemProps.key = 'body';
        itemProps.id = `boldrui-accordion__body-${itemUuid}`;
        itemProps.role = accordion ? 'tabpanel' : '';
      }

      return React.cloneElement(item, itemProps);
    });
  }

  render() {
    return (
      <div className="boldrui-accordion__item">
        {this.renderChildren()}
      </div>
    );
  }
}

export default AccordionItem;
