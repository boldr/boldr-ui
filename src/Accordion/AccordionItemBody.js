/* @flow */
import React from 'react';
import type { Node } from 'react';
import classNames from 'classnames';

type Props = {
  id: string,
  isExpanded: boolean,
  children: Array<Node>,
  className: string,
  hideBodyClassName: string,
  role: string,
};

const defaultProps = {
  isExpanded: false,
  className: 'boldr-accordion__body',
  hideBodyClassName: 'boldr-accordion__body--hidden',
};

const AccordionItemBody = ({
  id,
  isExpanded,
  children,
  className,
  hideBodyClassName,
  role,
}: Props) => {
  const bodyClass = classNames(className, {
    [hideBodyClassName]: !isExpanded,
  });
  const ariaHidden = !isExpanded;
  return (
    <div id={id} className={bodyClass} aria-hidden={ariaHidden} role={role}>
      {children}
    </div>
  );
};

AccordionItemBody.defaultProps = defaultProps;
AccordionItemBody.accordionElementName = 'AccordionItemBody';

export default AccordionItemBody;
