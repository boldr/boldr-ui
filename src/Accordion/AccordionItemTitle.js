/* @flow */
import React from 'react';
import type { Node } from 'react';
import cn from 'classnames';

type Props = {
  id: string,
  isExpanded: boolean,
  onClick: Function,
  ariaControls?: string,
  children: Array<Node>,
  role: string,
  className?: string,
  hideBodyClassName?: string,
};

class AccordionItemTitle extends React.Component<Props, void> {
  static defaultProps = {
    id: '',
    isExpanded: false,
    onClick: () => {},
    ariaControls: '',
    role: '',
  };

  handleKeyPress = (e: Event) => {
    const { onClick } = this.props;
    if (e.charCode === 13 || e.charCode === 32) {
      onClick();
    }
  };
  props: Props;
  render() {
    const {
      id,
      isExpanded,
      ariaControls,
      onClick,
      children,
      className,
      role,
      hideBodyClassName,
    } = this.props;
    const titleClassName = cn('boldrui-accordion__title', className, {
      [hideBodyClassName]: hideBodyClassName && !isExpanded,
    });

    return (
      <div // eslint-disable-line jsx-a11y/no-static-element-interactions
        id={id}
        aria-expanded={isExpanded}
        aria-controls={ariaControls}
        className={titleClassName}
        onClick={onClick}
        role={role}
        tabIndex="0"
        onKeyPress={this.handleKeyPress}
      >
        {children}
      </div>
    );
  }
}

export default AccordionItemTitle;
