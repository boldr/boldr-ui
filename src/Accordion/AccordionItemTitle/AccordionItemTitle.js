/* @flow */
import React, { PureComponent } from 'react';

type Props = {
  id: string,
  expanded: boolean,
  onClick: () => void,
  ariaControls: string,
  children: ReactChildren,
  role: string,
};

class AccordionItemTitle extends PureComponent {
  static defaultProps = {
    id: '',
    expanded: false,
    onClick: () => {},
    ariaControls: '',
    role: '',
  };

  handleKeyPress = (e: Event) => {
    const { onClick } = this.props;
    if (e.charCode === 13 || e.charCode === 32) {
      onClick();
    }
  };
  props: Props;
  render() {
    const { id, expanded, ariaControls, onClick, children, role } = this.props;
    return (
      <div // eslint-disable-line jsx-a11y/no-static-element-interactions
        id={id}
        aria-expanded={expanded}
        aria-controls={ariaControls}
        className="boldrui-accordion__title"
        onClick={onClick}
        role={role}
        onKeyPress={this.handleKeyPress}
      >
        {children}
      </div>
    );
  }
}
AccordionItemTitle.accordionElementName = 'AccordionItemTitle';

export default AccordionItemTitle;
