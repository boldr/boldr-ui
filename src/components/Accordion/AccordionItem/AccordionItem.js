import React, { Component } from 'react';
import uuid from 'uuid';
import type { ReactChildren } from '../../../types/react.js.flow';

type Props = {
  expanded: boolean,
  accordion: boolean,
  onClick: () => void,
  children: ReactChildren,
  className: string,
};

class AccordionItem extends Component {
  static defaultProps = {
    accordion: true,
    expanded: false,
    onClick: () => {},
    className: 'boldrui-accordion__item',
  };
  constructor(props) {
    super(props);
    this.state = {
      itemUuid: uuid.v4(),
    };
    this.renderChildren = this.renderChildren.bind(this);
  }
  props: Props;
  renderChildren() {
    const { accordion, expanded, onClick, children } = this.props;
    const { itemUuid } = this.state;

    return React.Children.map(children, item => {
      const itemProps = {
        expanded,
      };

      if (item.type.accordionElementName === 'AccordionItemTitle') {
        itemProps.key = 'title';
        itemProps.id = `boldrui-accordion__title-${itemUuid}`;
        itemProps.ariaControls = `boldrui-accordion__body-${itemUuid}`;
        itemProps.onClick = onClick;
        itemProps.role = accordion ? 'tab' : 'button';
      } else if (item.type.accordionElementName === 'AccordionItemBody') {
        itemProps.key = 'body';
        itemProps.id = `boldrui-accordion__body-${itemUuid}`;
        itemProps.role = accordion ? 'tabpanel' : '';
      }

      return React.cloneElement(item, itemProps);
    });
  }

  render() {
    const { className } = this.props;
    return (
      <div className={className}>
        {this.renderChildren()}
      </div>
    );
  }
}

export default AccordionItem;
