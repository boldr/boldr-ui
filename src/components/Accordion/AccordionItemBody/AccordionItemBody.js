/* @flow */
import React from 'react';
import classNames from 'classnames';
import type { ReactChildren } from '../../../types/react.js.flow';

type Props = {
  id: string,
  expanded: boolean,
  onClick: () => void,
  ariaControls: string,
  children: ReactChildren,
  className: string,
  hideBodyClassName: string,
  role: string,
};

const defaultProps = {
  id: '',
  expanded: false,
  className: 'boldrui-accordion__body',
  hideBodyClassName: 'boldrui-accordion__body--hidden',
  role: '',
};

const AccordionItemBody = ({
  id,
  expanded,
  children,
  className,
  hideBodyClassName,
  role,
}: Props) => {
  const bodyClass = classNames(className, {
    [hideBodyClassName]: !expanded,
  });
  const ariaHidden = !expanded;
  return (
    <div id={id} className={bodyClass} aria-hidden={ariaHidden} role={role}>
      {children}
    </div>
  );
};

AccordionItemBody.defaultProps = defaultProps;
AccordionItemBody.accordionElementName = 'AccordionItemBody';

export default AccordionItemBody;
