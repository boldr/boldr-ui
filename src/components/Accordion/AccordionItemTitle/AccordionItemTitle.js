/* @flow */
import React, { PureComponent } from 'react';
import type { ReactChildren } from '../../../types/react.js.flow';

type Props = {
  id: string,
  expanded: boolean,
  onClick: () => void,
  ariaControls: string,
  children: ReactChildren,
  className: string,
  role: string,
};

class AccordionItemTitle extends PureComponent {
  static defaultProps = {
    id: '',
    expanded: false,
    onClick: () => {},
    ariaControls: '',
    className: 'boldrui-accordion__title',
    role: '',
  };

  handleKeyPress = (e: Event) => {
    const { onClick } = this.props;
    if (e.charCode === 13 || e.charCode === 32) {
      onClick();
    }
  };
  props: Props;
  render() {
    const {
      id,
      expanded,
      ariaControls,
      onClick,
      children,
      className,
      role,
    } = this.props;
    return (
      <div // eslint-disable-line jsx-a11y/no-static-element-interactions
        id={id}
        aria-expanded={expanded}
        aria-controls={ariaControls}
        className={className}
        onClick={onClick}
        role={role}
        tabIndex="0"
        onKeyPress={this.handleKeyPress}
      >
        {children}
      </div>
    );
  }
}
AccordionItemTitle.accordionElementName = 'AccordionItemTitle';

export default AccordionItemTitle;
