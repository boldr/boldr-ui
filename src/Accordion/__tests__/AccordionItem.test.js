import React from 'react';
import renderer from 'react-test-renderer';

import AccordionItemTitle from '../AccordionItemTitle';
import AccordionItemBody from '../AccordionItemBody';
import AccordionItem from '../AccordionItem';

jest.mock('../AccordionItemTitle', () => {
  const RealReact = require.requireActual('react');
  class MockElement extends RealReact.Component {
    render() {
      return RealReact.createElement('div', this.props, this.props.children);
    }
  }
  MockElement.accordionElementName = 'AccordionItemTitle';

  return MockElement;
});

jest.mock('../AccordionItemBody', () => {
  const RealReact = require.requireActual('react');
  class MockElement extends RealReact.Component {
    render() {
      return RealReact.createElement('div', this.props, this.props.children);
    }
  }
  MockElement.accordionElementName = 'AccordionItemBody';

  return MockElement;
});

describe('AccordionItem', () => {
  it('renders correctly with accordion true', () => {
    const tree = renderer
      .create(
        <AccordionItem>
          <AccordionItemTitle>
            <div>Fake title</div>
          </AccordionItemTitle>
          <AccordionItemBody>
            <div>Fake body</div>
          </AccordionItemBody>
        </AccordionItem>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with accordion false', () => {
    const tree = renderer
      .create(
        <AccordionItem isAccordion={false}>
          <AccordionItemTitle>
            <div>Fake title</div>
          </AccordionItemTitle>
          <AccordionItemBody>
            <div>Fake body</div>
          </AccordionItemBody>
        </AccordionItem>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('still renders with incorrect type children', () => {
    const tree = renderer
      .create(
        <AccordionItem isAccordion={false}>
          <div>Fake title</div>
          <div>Fake body</div>
        </AccordionItem>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with different className', () => {
    const tree = renderer
      .create(
        <AccordionItem className="testCSSClass">
          <AccordionItemTitle>
            <div>Fake title</div>
          </AccordionItemTitle>
          <AccordionItemBody>
            <div>Fake body</div>
          </AccordionItemBody>
        </AccordionItem>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
