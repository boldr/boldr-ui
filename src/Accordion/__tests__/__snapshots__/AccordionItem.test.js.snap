// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem renders correctly with accordion false 1`] = `
<div
  className="boldr-accordion__item"
>
  <div
    ariaControls="boldr-accordion__body-2"
    id="boldr-accordion__title-2"
    isExpanded={false}
    onClick={[Function]}
    role="button"
  >
    <div>
      Fake title
    </div>
  </div>
  <div
    id="boldr-accordion__body-2"
    isExpanded={false}
    role=""
  >
    <div>
      Fake body
    </div>
  </div>
</div>
`;

exports[`AccordionItem renders correctly with accordion true 1`] = `
<div
  className="boldr-accordion__item"
>
  <div
    ariaControls="boldr-accordion__body-1"
    id="boldr-accordion__title-1"
    isExpanded={false}
    onClick={[Function]}
    role="tab"
  >
    <div>
      Fake title
    </div>
  </div>
  <div
    id="boldr-accordion__body-1"
    isExpanded={false}
    role="tabpanel"
  >
    <div>
      Fake body
    </div>
  </div>
</div>
`;

exports[`AccordionItem renders with different className 1`] = `
<div
  className="boldr-accordion__item"
>
  <div
    ariaControls="boldr-accordion__body-4"
    id="boldr-accordion__title-4"
    isExpanded={false}
    onClick={[Function]}
    role="tab"
  >
    <div>
      Fake title
    </div>
  </div>
  <div
    id="boldr-accordion__body-4"
    isExpanded={false}
    role="tabpanel"
  >
    <div>
      Fake body
    </div>
  </div>
</div>
`;

exports[`AccordionItem still renders with incorrect type children 1`] = `
<div
  className="boldr-accordion__item"
>
  <div>
    Fake title
  </div>
  <div>
    Fake body
  </div>
</div>
`;
